Globe - basic infrastructure, raster layer adaptation for globe

kubapelc opened the pull request:
This is the first PR of the implementation of vector globe view for MapLibre. It is a selection of changes from my main [vector globe branch](https://github.com/windycom/maplibre-gl-js/tree/kubapelc/globe-vector) as well as interface and changes from @pheonor's globe implementation. More PRs will follow with more of the globe rendering implementation. I plan to split PRs into smaller ones for easier reviewing, as requested.

Issue & discussion of globe:

https://github.com/maplibre/maplibre/issues/190
https://github.com/maplibre/maplibre/discussions/161

## Additions and changes

- base classes and interface for globe, a unification of both my and @pheonor's forks, so that we can both do PRs independently in the future
- changes to `map.ts` and Map constructor to include projection
- system for projection-dependent vertex shader prelude code
	- such code is injected at the beginning of every vertex shader
	- see `src/shaders/_projection_mercator.vertex.glsl` for an example
	- example: function `projectTile()` projects a point in local tile coordinates (0..EXTENT) to screen
		- for mercator, this is mostly equivalent to a simple matrix multiplication (usually replaces `u_matrix * vec4(a_pos, 0.0, 1.0)` in shaders)
		- for globe, this does a projection from mercator to a 3D point on the surface of a sphere, which is then transformed with a projection matrix
		- additionally, for globe projection, this function automatically handles animated transition from globe to mercator (for high zoom levels)
	- this system should (in theory) allow for simpler implementation of different projections in the future
- `ProjectionBase` class, and its implementations for Mercator and Globe projections
	- note that globe projection uses an instance of mercator projection internally, since globe transitions to mercator at high zoom levels
	- the currently used projection class is exposed in `map.ts` to allow custom layers to implement globe view using shader projection preludes
	- the projection class will likely expand in the future once proper controls and other non rendering-related features are implemented for globe
- adaptation of the raster layer for globe (other layers are still mercator)
 	- raster layer is a relatively simple example of how the new projection classes and projection shader preludes are used
- an example webpage with globe view (`test/examples/globe.html`)
- some minor changes
	- additional comments
	- typo fixes
	- simplify `tilesAfterTime` implementation in `terrain_source_cache.ts`

## Example

![A screenshot of a globe with a satellite map](https://github.com/maplibre/maplibre-gl-js/assets/57600346/32b21417-63f4-41af-b3a1-d16a72308e81)

See `test/examples/globe.html` for a web page that displays a satellite map projected onto a globe. Note that only the raster layer is currently adapted for globe projection, so lines and symbols will still display as if on a mercator map.

The projection will automatically transition to mercator at high zoom levels. The transition is animated and happens at a point when globe is already almost indistinguishable from mercator projection, so it is nearly unnoticeable.

## Changes compared to Pheonor's repo

@pheonor please also review this PR, especially the following changes and adaptations from your branch:

- I've excluded the `project` and `unproject` methods from the projection classes for now, as they are not used by anything (yet).
- I've moved the place where projection is stored from `transform.ts` into `map.ts`, as transform might exist in multiple instances, but map should (as far as I know) only ever exist in one single instance.
- The globe example webpage is from your branch, just slightly modified (added a title and description).

## Problem: where to place granularity settings?

Granularity determines how much subdivision is applied to tile geometry. For example, a subdivision granularity of 4 will "split" all geometry in a tile as if it was cut by the edges of a 4x4 grid. Different zoom levels have different granularities. Granularity for a given zoom level is determined by a simple function.

It is relatively clear how to set granularity levels for globe projection, and thus, if MapLibre only ever used globe view, those values could be static. But what if the user wants a map that will always use mercator? Then there is no need for subdivision to slow down tile parsing and rendering. Or what if a different projection in the future needs different granularities?

Subdivision functions are used in `*_bucket.ts` (fill, fill-extrusion and line buckets), right before generating vertex buffers. The bucket functions run in a separate web worker which has no access to the map or transform instance, as far as I know. Furthermore, the granularity doesn't depend *only* on zoom level, but also on what type of geometry is getting subdivided - for example line layers use higher granularity than fill layers.

Currently, granularity settings are static, both in the sense that they cannot change eg. with projection and that they use the `static` keyword to be accessible from anywhere.

I have no idea how to solve this. The data about which tiles need to be loaded (data which ultimately ends up in the bucket) passes through complex machinery and it is hard for me to even navigate through it.

In this PR, subdivision is only used for stencil masks, but following PRs will utilize it for fill, fill extrusion and line geometry as well.

## Failing tests & other issues

As far as I know, about 300 render tests fail in my branch. After consulting with @wipfli I decided to publish the PR anyway, but I will continue debugging the tests and I'll try to find a fix. I don't know about the state of unit tests, I haven't managed to even run those yet. But using this branch of MapLibre for displaying a map in a web browser should work without any issues.

~~On Chrome, you may notice a warning about a GPU buffer guarded by a FenceSync object being read before the fence is signaled. I know about this issue and will attempt to fix it. It is related to globe projection (see class `ProjectionErrorMeasurement` in `src/geo/projection/globe.ts`).~~ This is now fixed.

## Checklist

- [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
- [x] Briefly describe the changes in this PR.
- [x] Link to related issues.
- [x] Include before/after visuals or gifs if this PR includes visual changes.
- [ ] Write tests for all new functionality.
- [x] Document any changes to public APIs.
	- Completed, as long as doc comments are enough.
- [ ] Post benchmark scores.
- [ ] Add an entry to `CHANGELOG.md` under the `## main` section.

HarelM commented:
Thanks a lot for opening this PR!!
It will take me some time to review it as it has a lot of changes.
I'll start with some basic stuff and then later on move to the more in depth, so please bare with me.
Also note that you can use this PR to see the results of the render tests.
If your code is "opt in" then the render tests should not fail, in theory, but everything is possible.
If you have any issue running the tests or understanding what they do, let me know and I'll do my best to help.
It seems that the unit tests passed, so this is a good sign.

HarelM commented on src/geo/projection/projection_base.ts:
```diff
@@ -0,0 +1,125 @@
+/**
+ * An abstract class the specializations of which are used internally by MapLibre to handle different projections.
+ */
+export abstract class ProjectionBase {
```
Can this be converted to an interface? I prefer composition over inheritance and also this class seems to have very little implementation.

kubapelc replied:
Yes, this can absolutely be an interface, I don't think ProjectionBase should ever contain any nontrivial logic. Converted to interface in the latest commit.

kubapelc commented:
I fixed the issue with Chrome warning related to FenceSync and pixel readback.

HarelM commented:
@kubapelc do you think you can create a PR only for the `Mesh` class addition and changes? (should be like 4-5 files I think).
I want to push only this and try and use this `Mesh` class in the sky maybe?
I can try and do it if you find it complicated...

HarelM commented:
For now I copied the `mesh.ts` file to the sky2 branch, I hope it won't cause a merge issue once sky is merge...

HarelM commented on src/ui/map.ts:
````diff
@@ -3324,4 +3362,24 @@ export class Map extends Camera {
+     * let projectionName = map.getProjectionName();
+     * ```
+     */
+    getProjectionName(): string { return this.projection.name; }
````
I'm not sure this is an interesting API since we have `getProjection` and one can get the name from it.

kubapelc replied:
True, I removed it.

Pheonor replied:
When I add this API, I want a symetric API to set projection with just the name and get the projection name we use just before. But I also think that getting the full projection object could be usefull.

kubapelc replied:
I think that setting a projection at runtime should not be done with a setter, but rather with a function, since it will probably be a quite heavy operation. Something like `setProjectionByName(...)`.

HarelM replied:
Sure, I agree.

HarelM replied:
P.S. here's a meduim article I wrote:
https://medium.com/@harel.mazor/why-i-never-use-get-and-set-in-any-code-language-7a0f7de739ce

Pheonor replied:
Sure, this is exactly I propose into my branch.
```
    setProjection(projectionName: string): Map {
        this.transform.projection = projectionName;

        this.updateRenderToTexture(this.getTerrain());

        this.fire(new Event('projection', {projection: projectionName}));
        return this._update();
    }
```
We could add it later, in another PR.

HarelM commented on src/ui/map.ts:
```diff
@@ -58,9 +58,32 @@ import type {
+ */
+type ProjectionName = 'mercator' | 'globe';
+
+function getProjectionFromName(name: ProjectionName, map: Map): ProjectionBase {
```
To avoid all kind of circular dependencies, I would recommend placing this method in a file of its own, it will probably grow in the future too.

kubapelc replied:
I've moved it to a separate file.

HarelM commented on src/ui/map.ts:
````diff
@@ -3324,4 +3362,24 @@ export class Map extends Camera {
+     * let projection = map.getProjection();
+     * ```
+     */
+    getProjection(): ProjectionBase { return this.projection; }
````
What about changing the projection in terms of API?

kubapelc replied:
I'm not sure whether to allow projection to be changed at runtime. See the section on granularity settings in the PR description. Tl;dr globe needs to subdivide almost all geometry during the loading stage, which slows down loading of new tiles and maybe also slows down rendering a bit due to drawing more polygons.

Right now, this subdivision happens regardless of what projection is used, but I think that it would be a really good idea in the future to allow some projections (mercator mostly) to disable subdivision altogether to boost performance.

Problem is that if you then wanted to switch from mercator to globe, you would need to reload all tiles to apply proper subdivision to them.

If there is a mechanism to force the reload of all tiles (or at least the regeneration of their vertex buffers) in MapLibre already, then runtime projection changing can be implemented with relative ease.

HarelM replied:
I'm not sure there is a mechanism to invalidate all the tiles, but I know this is needed, I've seen plugins that needed it and are using internal APIs related to source cache, which is a great way to break stuff (maplibre-inspect for example).
So generally speaking, I think we need an API to clear tiles...

Pheonor replied:
Effectively, it could be useful to change projection at runtime. We just need to reset all cache and render all tiles.

kubapelc replied:
I'd like to go ahead without runtime projection switching at the moment, and add that (plus proper handling of subdivision) in a later PR. My goal is to have this PR merged rather sooner than later, so that I can polish and create PRs of other globe features. The reviewing process for PRs will necessarily have a lot of latency, so I can work on projection switching in the meantime.

HarelM replied:
Sounds like a good approach.

HarelM commented on src/source/source_cache.ts:
```diff
@@ -950,8 +950,23 @@ export class SourceCache extends Evented {
 
-    getVisibleCoordinates(symbolLayer?: boolean): Array<OverscaledTileID> {
-        const coords = this.getRenderableIds(symbolLayer).map((id) => this._tiles[id].tileID);
+    getVisibleCoordinates(symbolLayer?: boolean, deduplicateWrapped?: boolean): Array<OverscaledTileID> {
```
It is not clear what `deduplicateWrapped` flag is for, can you elaborate?

HarelM replied:
Probably worth adding a unit test to this case as well as the method is not trivial.

kubapelc replied:
I've added a comment that should clarify it:

```ts
getVisibleCoordinates(symbolLayer?: boolean, deduplicateWrapped?: boolean): Array<OverscaledTileID> {
        let coords = this.getRenderableIds(symbolLayer).map((id) => this._tiles[id].tileID);

        // When rendering a mercator map, the screen may be larger than the world,
        // causing multiple "copies" of the world to be visible and
        // the same tile might be drawn in two different places (see wrap in OverscaledTileID).
        // For globe, we only ever want to draw a single copy of the world, so no (canonical) tile should be present
        // multiple times in the resulting list of tile IDs.
        if (deduplicateWrapped) {
```

I'm not 100% confident about what *exactly* the method is doing and why `symbolLayer` flag exists to write a proper doc comment for it.

HarelM replied:
If this is the case, changing it sounds risky... I would consider trying to better understand what it does before changing it, once you are sure about what it does, writing a comment should be an easy task.

kubapelc replied:
I've taken a deeper look at how visible tiles are determined, and also at what `renderWorldCopies` flag does, as you suggested in another thread. You are right that the flag does something similar to what I am trying to do with `deduplicateWrapped`, but it also has some side effects (constrains zoom and map center to only show one world copy at any time).

It might be better to implement the tile deduplication in `transform.ts`, in the `coveringTiles` method. But, additionally, proper determination of what tiles are visible on a globe map is actually part of the same problem. I haven't even started work on that, but @Pheonor has already implemented it in his branch, see changes around https://github.com/Pheonor/maplibre-gl-js/blob/300f8303c4b4e2df03b9d1773e42e4b2f8e4baea/src/geo/transform.ts#L498

Right now I see two options: I can temporarily move `deduplicateWrapped` functionality to `transform.ts`, or I can just remove it altogether for now. The reason is that the entire `coveringTiles` will need to be changed in the near future to accomodate different projections, and probably moved from `transform.ts` to the projection classes. So I see no point in coming up with a nice solution at this moment.

HarelM replied:
If it doesn't break anything and we are about to replace it, let's remove it.

kubapelc replied:
Done, removed.

HarelM commented on src/source/image_source.ts:
```diff
@@ -104,6 +105,7 @@ export class ImageSource extends Evented implements Source {
     _boundsArray: RasterBoundsArray;
     boundsBuffer: VertexBuffer;
     boundsSegments: SegmentVector;
+    tileCoords: Array<Point>;
```
Can you elaborate on this change?
There's this issue that is worth considering if imagesource is touched I guess:
https://github.com/maplibre/maplibre-gl-js/issues/1559

kubapelc replied:
Images used to be rendered with a quad that used a vertex buffer with custom positions (covering the rectagle needed by a given ImageSource) and texcoords that were always the corners of the 0..1 range. Regular raster tiles were drawn in the same way, just the vertex positions were always 0..EXTENT range corners.

For globe I needed to draw both raster tiles and custom images with subdivided geometry. I didn't want to generate a new subdivided grid mesh for each custom image, si I did the following:

Every image/raster tile uses a mesh with vertex positions in range 0..EXTENT. I can cache and reuse those meshes for a given level of subdivision, so only about a dozen of those will ever exist. Images may need to have different corner positions than 0,0 or EXTENT,EXTENT but I already need to pass the actual extents of the tile to the shader for globe (see code around [this line](https://github.com/windycom/maplibre-gl-js/blob/3cc71ca392f71d16b5414d3cf82fd7f27679a124/src/render/draw_raster.ts#L104) ), so I can infer the correct corner coordinates from that in the shader.

The line you highlighed exposes the corner coordinates for image sources, so that I can access them during rendering and pass them to the shader.

Now that I know that the mesh will always be a grid in 0..EXTENT, I no longer need the texture coordinates in the vertex buffer, because I can compute them from vertex positions. So I removed the texture coordinates.

I've also noticed just now that the `boundsBuffer` and other vertex-buffer related members of ImageSource are now unused, so I will also remove them.

kubapelc replied:
Also I think this sadly has no relation to the issue you mentioned, that one seems to be specific to terrain rendering.

HarelM replied:
OK, thanks for the info.

kubapelc commented:
> For now I copied the `mesh.ts` file to the sky2 branch, I hope it won't cause a merge issue once sky is merge...

I think just copying the class is ok, it should not cause any difficult merge issues. But I can still create the separate PR if you want.

HarelM commented on test/examples/globe.html:
```diff
@@ -0,0 +1,28 @@
+<script>
+    const map = new maplibregl.Map({
+        container: 'map',
+        projection: 'globe',
```
This should be part of the style itself, or if the spec is still not up-to-date, set using a map API call like "setProjection".

HarelM replied:
Might not be relevant, see my comments:
https://github.com/maplibre/maplibre-gl-js/pull/3783#issuecomment-1996770425

HarelM commented on src/ui/map.ts:
```diff
@@ -318,6 +320,13 @@ export type MapOptions = {
+     * - 'globe' - A 3D spherical view of the planet when zoomed out, transitioning seamlessly to Web Mercator at high zoom levels.
+     * @defaultValue 'mercator'
+     */
+    projection?: ProjectionName;
```
See my other comments about where this definition should be.

HarelM commented on src/source/video_source.ts:
```diff
@@ -161,14 +159,6 @@ export class VideoSource extends ImageSource {
         const context = this.map.painter.context;
         const gl = context.gl;
 
-        if (!this.boundsBuffer) {
```
Can you elaborate on why this was removed?

kubapelc replied:
I changed the implementation of rendering raster data (apart from raster layers this also includes video, images and canvas), now it no longer needs custom vertex buffers. `boundsBuffer` was thus removed from the base class `ImageSource`. See my comment at https://github.com/maplibre/maplibre-gl-js/pull/3783#discussion_r1518571524

HarelM replied:
OK, thanks!

HarelM commented on src/source/source_cache.ts:
```diff
@@ -950,8 +950,28 @@ export class SourceCache extends Evented {
+        // causing multiple "copies" of the world to be visible and
+        // the same tile might be drawn in two different places (see wrap in OverscaledTileID).
+        // For globe, we only ever want to draw a single copy of the world, so no (canonical) tile should be present
+        // multiple times in the resulting list of tile IDs.
```
There is the renderWorldCopies flag, sounds like it does similar behavior, can we use it instead of deduplicating things?

kubapelc replied:
This is related to the conversation about `deduplicateWrapped`, please see https://github.com/maplibre/maplibre-gl-js/pull/3783#discussion_r1523424897

HarelM commented on src/shaders/_projection_mercator.vertex.glsl:
```diff
@@ -0,0 +1,24 @@
+float projectLineThickness(float tileY) {
```
Why does this file name start with "_"?

kubapelc replied:
It starts with "_" to stay consistent with the naming of `_prelude.vertex.glsl` and `_prelude.fragment.glsl`, since this file also contains a "prelude" that is placed at the beginning of every shader. Do you think removing the "_" or moving it to a subdirectory would be better?

HarelM replied:
If this server the same purpose of _prelude in terms of being placed at the beginning of each shader why not add it to the prelude file?
While I don't like the idea of having a large file with a lot of methods, it might server the purpose.
Or we can call it `_prelude_for_projections`?

kubapelc replied:
Either `_projection_mercator.vertex.glsl` or `_projection_globe.vertex.glsl` gets inserted at the start of all shaders based on what projection is used, each file has a projection-dependet implementation of the shader functions. I used to have both in `_prelude.vertex.glsl` and used `#ifdef GLOBE` to switch, but I figured placing them into different files is a nicer approach, especially since there might be more than two projections in the future.

HarelM replied:
I see. Your call, the "-" prefix is OK.

HarelM commented on src/render/subdivision.ts:
```diff
@@ -0,0 +1,52 @@
+}
+
+export const granularitySettings: SubdivisionGranularitySetting = new SubdivisionGranularitySetting(
+    new SubdivisionGranularityExpression(7, 1), // Fill
```
These are hard coded, I'm guessing they were chosen empirically, consider allowing these to be changed somehow if this wasn't already done.

kubapelc replied:
Yes, these values are for globe projection, a sweetspot between producing a sufficiently tesselated/round globe while not generating too much extra geometry. I want to keep them readonly, since they cannot change once a map is loaded. See conversation around https://github.com/maplibre/maplibre-gl-js/pull/3783#discussion_r1518572684 - once runtime projection switching is implemented, every projection would have its own set of granularity settings. But they should still be constant for a given projection.

HarelM commented on src/render/program.ts:
```diff
@@ -75,12 +74,15 @@ export class Program<Us extends UniformBindings> {
+        if (hasTerrain) {
             defines.push('#define TERRAIN3D;');
         }
+        if (programDefines) {
```
Isn't this a bit too general? I think being explicit here would be better.
I also saw some `#ifndef GLOBE` that I guess should be added here...?

kubapelc replied:
I've refactored this constructor a bit.

HarelM commented on src/render/painter.ts:
```diff
@@ -566,20 +617,42 @@ export class Painter {
-                this.style.map.terrain
+                useTerrain,
+                projectionPrelude,
+                concatenatedDefines
```
See my previous comment.

HarelM commented on src/render/painter.ts:
```diff
@@ -566,20 +617,42 @@ export class Painter {
             (this._showOverdrawInspector ? '/overdraw' : '') +
-            (this.style.map.terrain ? '/terrain' : '');
+            (useTerrain ? '/terrain' : '') +
+            (concatenatedDefines ? (`/defines:${concatenatedDefines.join('//')}`) : '');
```
This seems to be getting pretty complicated, can this be simplified?

kubapelc replied:
I've refactored it a bit. Not necessarily to make it simpler, but at least easier to read I hope.

HarelM commented on src/render/painter.ts:
```diff
@@ -507,36 +577,17 @@ export class Painter {
-        const translatedMatrix = new Float32Array(16);
-        mat4.translate(translatedMatrix, matrix, translation);
-        return translatedMatrix;
+    // Temporary function - translate & translate-anchor handling will be moved to projection classes,
```
Please use `/**` type of comments above methods, it's better for the docs.
If this is to be moved to a different place you can ignore this comment assuming this will be part of this PR.

kubapelc replied:
I've changed it to `/**`.

HarelM commented on src/render/painter.ts:
```diff
@@ -217,9 +232,13 @@ export class Painter {
         mat4.scale(matrix, matrix, [gl.drawingBufferWidth, gl.drawingBufferHeight, 0]);
 
-        this.useProgram('clippingMask').draw(context, gl.TRIANGLES,
+        const projectionData = this.style.map.projection.getProjectionData(null, null);
```
Why are we sending `null, null` to the method? should these be optional parameters?

kubapelc replied:
This is a bit of a special case, the method just draws a fullscreen quad, so I force a simple (mercator) shader that only uses a matrix multiplication for projection, and pass that matrix as an input instead of full `projectionData`. But `getProjectionData` should not have its arguments marked as optional I think, since it never makes sense to not use them - other than in this place.

I've refactored it so that instead of calling `getProjectionData(null, null)`, the mostly empty `ProjectionData` object is just created directly instead.

HarelM commented on src/render/painter.ts:
```diff
@@ -244,16 +263,27 @@ export class Painter {
 
+            let mesh = this.tileExtentMesh;
+
+            if (projection instanceof GlobeProjection) {
```
I usually don't like to use `instanceof` since this will require painter to be familiar with the implementation of globe project and may cause circular dependencies, consider using some interface proerty like `name` or `type`.

kubapelc replied:
I've noticed that most places where I need globe-specific codepaths are not actually globe-specific, but rather specific to any projection which uses subdivision, so I refactored the interface to avoid `instanceof GlobeProjection` checks.

HarelM commented on src/render/painter.ts:
```diff
@@ -356,15 +421,17 @@ export class Painter {
         const coordsDescending: {[_: string]: Array<OverscaledTileID>} = {};
         const coordsDescendingSymbol: {[_: string]: Array<OverscaledTileID>} = {};
 
+        const deduplicateWrapped = !style.map.projection.drawWrappedTiles;
```
As I said in previous comment, consider using renderWorldCopies.

HarelM commented on src/render/painter.ts:
```diff
@@ -408,6 +475,9 @@ export class Painter {
         }
 
+        // Execute offscreen GPU tasks of the projection manager
+        this.style.map.projection.updateGPUdependent(this);
```
Passing `this` as a parameter is usually a code smell and might need a redesign or rethink of what part is truely needed to be passed down.

kubapelc replied:
This is due to `ProjectionErrorMeasurement`. It mostly just needs the GL context, but it also needs to call `useProgram` on a Painter instance. I could just pass the function alongside the GL context, but that feels too specific for a generic projection interface.

HarelM replied:
It is better to be more specific.
Let's start with passing only what's needed, in the future if more parameters are needed we can pass those, this way you create less coupling.

kubapelc replied:
Fixed.

HarelM commented on src/render/draw_raster.ts:
```diff
@@ -22,58 +32,97 @@ export function drawRaster(painter: Painter, sourceCache: SourceCache, layer: Ra
     const program = painter.useProgram('raster');
 
+    const projection = painter.style.map.projection;
+    const globe = (projection instanceof GlobeProjection && projection.useGlobeRendering);
```
Same as other comment about instance of.

HarelM commented on src/render/draw_raster.ts:
```diff
@@ -22,58 +32,97 @@ export function drawRaster(painter: Painter, sourceCache: SourceCache, layer: Ra
 
-    const minTileZ = coords[coords.length - 1].overscaledZ;
+    // Stencil and two-pass is not used for ImageSource sources.
+    const passCount = (globe && !(source instanceof ImageSource)) ? 2 : 1;
```
There are too many if related to globe and image source, I would consider splitting this logic to different methods and call each one according to what is needed. It's almost impossible to understand what this method is doing...

kubapelc replied:
Yeah this is a mess 🤦 I've refactored it into something that is hopefully much more readable.

HarelM commented on src/geo/projection/projection_base.ts:
```diff
@@ -0,0 +1,120 @@
+     * @internal
+     * A short, descriptive name of this projection, such as 'mercator' or 'globe'.
+     */
+    get name(): string;
```
I generally don't like `get`ters and `set`ters. I prefer a method to get the relevant information.
https://medium.com/@harel.mazor/why-i-never-use-get-and-set-in-any-code-language-7a0f7de739ce

kubapelc replied:
I think a getter makes sense in this case, since its implementations should always just return a constant with no side effects.

HarelM commented on src/geo/projection/globe.ts:
```diff
@@ -0,0 +1,767 @@
+ */
+const EXTENT_STENCIL_BORDER = EXTENT / 128;
+
+function clamp(a: number, min: number, max: number): number {
```
Consider moving this to utils file.

kubapelc replied:
Done, I've also removed `smoothStep` and used `easeCubicInOut` instead.

HarelM commented on src/geo/projection/globe.ts:
```diff
@@ -0,0 +1,767 @@
+const errorTransitionTimeSeconds = 0.5;
+
+export class GlobeProjection implements ProjectionBase {
+    private _map: Map | undefined;
```
While I'm all for `private` and `public` stuff, this codebase doesn't use it at all, I don't know if we should introduce it. Your call...

kubapelc replied:
I'd prefer to use `private` and `public`, and it seems both keywords are already used in some places (`worker.ts`, `render_pool.ts`).

HarelM commented on src/geo/projection/globe.ts:
```diff
@@ -0,0 +1,767 @@
+import {granularitySettings} from '../../render/subdivision';
+import Point from '@mapbox/point-geometry';
+import {ProjectionData} from '../../render/program/projection_program';
+import * as Mercator from './mercator';
```
Why not import only the part we are interested from this file?

kubapelc replied:
Done.

HarelM commented on src/geo/projection/globe.ts:
```diff
@@ -0,0 +1,767 @@
+
+    private _globeProjectionOverride = true;
+
+    private _globeProjMatrix: mat4 = [
```
Isn't this a simple identity matrix? I think gl-matrix has methods for this...?

kubapelc replied:
I've replaced the initialization with `mat4.create();`

HarelM commented on src/geo/projection/globe.ts:
```diff
@@ -0,0 +1,767 @@
+    private _lastGlobeChangeTime: number = -1000.0;
+    private _lastLargeZoomStateChange: number = -1000.0;
+    private _lastLargeZoomState: boolean = false;
+    private _globeness: number = 1.0;
```
I'm not sure this is a great name, but I think a comment to better explain what it does can help.

kubapelc replied:
I've added a comment to `_globeness`.

HarelM commented on src/geo/projection/globe.ts:
```diff
@@ -0,0 +1,767 @@
+            cameraPos[2] / cameraPos[3]
+        ];
+
+        // We want to compute a plane equation that, when applied to the unit sphere generated
```
This comment probably indicates that this should be moved to a method, preferably with the below input variables.

kubapelc replied:
Yes this makes sense, I've moved it to a separate method.

HarelM commented on src/geo/projection/globe.ts:
```diff
@@ -0,0 +1,767 @@
+        this._fbo = context.createFramebuffer(this._texWidth, this._texHeight, false, false);
+        this._fbo.colorAttachment.set(texture);
+
+        if (this._allowWebGL2 && gl instanceof WebGL2RenderingContext) {
```
Is this error correction only relevant to webgl2? We are writing code to facilitate for both webgl1 and 2 in maplibre.

kubapelc replied:
It has a fallback for WebGL1. The `_allowWebGL2` variable is something I used to test this fallback, it is not needed anymore, so I'll remove it.

HarelM commented on src/geo/projection/globe.ts:
```diff
@@ -0,0 +1,767 @@
+    }
+}
+
+function parseRGBA8float(buffer: Uint8Array): number {
```
Same as above, consider moving this method to utils.

kubapelc replied:
This method is specific to how ProjectionErrorMeasurement encodes the value, so I'd prefer to leave it where it is. It is not a general utility function.

HarelM commented:
I have reviewed all the files, not sure my review is good enough though as my knowledge of GPU and shaders is very limited.
Would be nice if @Pheonor and maybe @JannikGM would review this as well.

Since this is targeting the globe branch, we can decide if we want to merge this sooner to facilitate more work on the globe branch.
Any feedback is welcome.

kubapelc commented on src/render/subdivision.ts:
```diff
@@ -0,0 +1,52 @@
+     */
+    public readonly granularityLine;
+
+    constructor(fill: SubdivisionGranularityExpression, line: SubdivisionGranularityExpression, stencil: SubdivisionGranularityExpression) {
```
Good point, I changed the constructor argument to an object.

HarelM commented on src/render/draw_raster.ts:
```diff
@@ -47,82 +40,86 @@ export function drawRaster(painter: Painter, sourceCache: SourceCache, layer: Ra
-    if (passCount > 1) {
-        [stencilModesHigh, stencilModesLow, coords] = painter.stencilConfigForOverlapTwoPass(tileIDs);
+    // Stencil mask and two-pass is not used for ImageSource sources regardless of projection.
+    if (source instanceof ImageSource) {
```
This looks a lot better! but I would still consider moving all the logic related to ImageSource to a separate method (that might call other small methods in this class) as there are still a few places in the draw tiles that checks if this is an image source or not.

kubapelc replied:
I've refactored it a bit more to remove `instanceof ImageSource` from `drawTiles`.

HarelM commented on src/geo/projection/globe_projection_error_measurement.ts:
```diff
@@ -0,0 +1,239 @@
+    }
+}
+
+function parseRGBA8float(buffer: Uint8Array): number {
```
Why not move it to be a private method of the above class? You can make it static if you want.

kubapelc replied:
Yes that makes sense, done.

HarelM commented on src/geo/projection/globe_projection_error_measurement.ts:
```diff
@@ -0,0 +1,239 @@
+    private _tryReadback(context: Context): void {
+        const gl = context.gl;
+
+        if (this._pbo && this._readbackQueue && gl instanceof WebGL2RenderingContext) {
```
I think we have `isWebGl2Supported` method somewhere, maybe we should use it?

kubapelc replied:
I've refactored it to use `isWebGL2`.

HarelM commented:
I think I resolved all the current implemented review comments, I left those that are still relevant I believe.
I've opened a design proposal for the style spec:
https://github.com/maplibre/maplibre-style-spec/issues/568
We might be able to ignore it for now like we do for the zoom and pitch that can be set for both the map configuration and the style, so it's not a blocker I guess, make sure to check my assumption.

kubapelc commented on test/integration/render/run_render_tests.ts:
```diff
@@ -56,6 +57,8 @@ type TestData = {
     diff: string;
     expected: string;
+
+    projection?: ProjectionName;
```
I've looked at `StyleSpecification`, but it seems to be an auto-generated file. I'm not sure about what I need to do to add projection into style spec.

HarelM replied:
There's a need to update the style spec package.
See my comment here: https://github.com/maplibre/maplibre-gl-js/pull/3783#issuecomment-1996770425
Let's just make sure this is similar to the zoom/center part of the map configuration.

HarelM commented:
I think most code review comments were resolved.
Please update the build test to have the updated bundle size so that all tests pass.
I'm wondering if more tests are needed or should we add them after more code changes have been made, as currently there are unit tests missing for the new classes, but this can wait for later PRs.

HarelM commented on src/render/painter.ts:
```diff
@@ -408,7 +474,17 @@ export class Painter {
         }
 
+        // Execute offscreen GPU tasks of the projection manager
+        const painterInstance = this;
```
`this` is captured when you create a lambda `() => {}`, so this assignment is not needed.

kubapelc replied:
Fixed.

HarelM commented on src/geo/projection/globe.ts:
```diff
@@ -0,0 +1,514 @@
+     */
+    get globeView(): boolean { return this._globeProjectionOverride; }
+    set globeView(value: boolean) {
+        if (value !== this._globeProjectionOverride) {
```
What's the point of this if?
Generally speking, I would simply make `_globeProjectionOverride` public, these getter and setter do not add anything I believe.

kubapelc replied:
Setting this used to have a side effect of updating the map (yeah, this was the bad kind of setter). I've now changed both setter and getter to be actual functions and refactored things a bit.

HarelM commented on src/ui/map.ts:
```diff
@@ -58,6 +58,8 @@ import type {
 import type {MapGeoJSONFeature} from '../util/vectortile_to_geojson';
 import type {ControlPosition, IControl} from './control/control';
 import type {QueryRenderedFeaturesOptions, QuerySourceFeatureOptions} from '../source/query_features';
+import {ProjectionBase} from '../geo/projection/projection_base';
```
Consider calling the projection interface simply "projection" and avoid the base part of the word.

kubapelc replied:
I wanted to avoid confusion with `symbol/projection.ts`, but as that file doesn't contain a class I think it's ok to go ahead with your suggestion.

kubapelc commented on src/geo/projection/globe.ts:
```diff
@@ -0,0 +1,514 @@
+        return this.useGlobeRendering;
+    }
+
+    get isRenderingDirty(): boolean {
```
I've changed this to a function.

kubapelc commented:
@HarelM I've added some render tests in a previous commit, and now I've added unit tests for globe and mercator as well, although they don't test too much functionality. I'll think about what else I can cover with unit tests.

I've fixed the bundle size in build test.

HarelM commented:
This is now merged to globe branch, I'll do my best to keep this brach up-to-date with main.
Please continue pushing improvements to it.
